Enable thp(transparent huge pages) for buffer sizes >=2MB

snadampal opened the pull request:
The 2MB thp pages provide better allocation latencies compared to the standard 4KB pages. This change has shown substantial improvement for batch mode usecases where the tensor sizes are larger than 100MB.

Only enabled if THP_MEM_ALLOC_ENABLE environment variable is set.

Relanding https://github.com/pytorch/pytorch/pull/93888 with functionality disabled for Android

snadampal commented:
Hi @izaitsevfb  appreciate if you could tell me what the next step is here, any action on me to get this merged?

snadampal commented:
Can we please target this PR for Pytorch 2.2?

malfet commented on c10/core/impl/alloc_cpu.h:
```diff
@@ -4,6 +4,11 @@
 
 #include <cstddef>
 
+#ifdef __linux__
+#include <sys/mman.h>
+#include <unistd.h>
+#endif
```
Can you please move this one to alloc_cpu.cpp (i.e. include what you use, this header does not need those includes)

snadampal replied:
Hi @malfet , I have rebased the PR and also addressed this comment.

snadampal committed:
Enable thp(transparent huge pages) for buffer sizes >=2MB

The 2MB thp pages provide better allocation latencies compared
to the standard 4KB pages. This change has shown substantial
improvement for batch mode usecases where the tensor sizes
are larger than 100MB.

Only enabled if THP_MEM_ALLOC_ENABLE environment variable is set.

malfet commented on c10/core/impl/alloc_cpu.cpp:
```diff
@@ -45,6 +50,35 @@ void memset_junk(void* data, size_t num) {
   }
 }
 
+#ifdef __linux__
```
Android is Linux, but we want to hide those in case of an Android, don't we?
```suggestion
#if defined(__linux__) && !defined(__ANDROID__)
```

malfet committed:
Update c10/core/impl/alloc_cpu.cpp

malfet commented:
@pytorchbot merge
